File: leetcode3/leetcode3_1.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        string subS;
        string longestS = "";
        for(int i = 0; i < s.length(); i++){
            subS = "";
            for(int j = i; j < s.length(); j++){
                if(subS.find(s[j]) != std::string::npos){
                    //cout << "break" << endl;
                    break;
                }
                else{
                    subS = subS + s[j];
                }
            }
            //cout << subS.size() << endl;
            if(subS.size() > longestS.size()){
                longestS = subS;
            }
        }
        
        return longestS.size();
    }
};
//Runtime: 380 ms
//Memory Usage: 913.4 KB
//time: O(n^2)
//space: O(s + subs + longestS)

File: leetcode3/leetcode3_1.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length(), ans = 0;
        Map<Character, Integer> map = new HashMap<>(); // current index of character
        
        // try to extend the range [i, j]
        for (int j = 0, i = 0; j < n; j++) {
            if (map.containsKey(s.charAt(j))) {
                i = Math.max(map.get(s.charAt(j)), i);
            }
            
            ans = Math.max(ans, j - i + 1);
            map.put(s.charAt(j), j + 1);
        }
        
        return ans;
    }
}

// Runtime: 6 ms, faster than 64.18% of Java online submissions for Longest Substring Without Repeating Characters.
// Memory Usage: 38.9 MB, less than 81.48% of Java online submissions for Longest Substring Without Repeating Characters.
// time complexity: O(n)
// space complexity: O(n)
// Solution 3